### ISSUE TITLE
[question] VirtualBuildEnv in Conan 1.x

### ISSUE THREAD

**Issue by db4** [conan-io/conan#14123]
### What is your question?

I have an internal tool used by other recipes via `tool_requires`. Its recipe contains
```python
    def package_info(self):
        self.cpp_info.libdirs = []
        self.cpp_info.includedirs = []
        # my_build_tool is in 'bin' subdirectory so the default 'bindirs' value is OK
```

When I use it with
```python
    generators = "VirtualBuildEnv"
    ...
    def build_requirements(self):
        self.tool_requires("my_build_tool/1.0")

    def build():
        self.run("my_build_tool ...")
```
my_build_tool's directory is correctly added to PATH and build() succeeds. But if I add the same requirement
```python
    generators = "VirtualBuildEnv"
    ...
    def requirements(self):
        self.requires("my_build_tool/1.0")

    def build_requirements(self):
        self.tool_requires("my_build_tool/1.0")

    def build():
        self.run("my_build_tool ...")
```
my_build_tool's directory will disappear from PATH and build() will fail. Is it expected behavior? Conan 2.0 doesn't suffer from that.

I see that CCI recipes always update PATH explicitly:
```python
        # TODO: to remove in conan v2
        bin_path = os.path.join(self.package_folder, "bin")
        self.output.info(f"Appending PATH environment variable: {bin_path}")
        self.env_info.PATH.append(bin_path)
```
but thought it should not be necessary as VirtualBuildEnv did it itself for cpp_info.bindirs.

### Have you read the CONTRIBUTING guide?

- [ ] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @db4 

Quick question: in the Conan 1.X case, are you providing the 2 profiles or just 1 profile? I suspect that maybe only 1? and if that is the case, can you please try with ``-pr:b=default``?

**Comment by db4** [comment]
Hi @memsharded

I've tried with and without `-pr:b=default` - the result is the same, bindirs are not added to PATH.

**Comment by memsharded** [comment]
I am not sure what it could be. I am trying to reproduce with:
```python
openssl = textwrap.dedent(r"""
        import os
        from conans import ConanFile
        from conans.tools import save, chdir
        class Pkg(ConanFile):
            settings = "os"
            def package(self):
                with chdir(self.package_folder):
                    echo = "@echo off\necho MYOPENSSL={}!!".format(self.settings.os)
                    save("bin/myopenssl.bat", echo)
                    save("bin/myopenssl.sh", echo)
                    os.chmod("bin/myopenssl.sh", 0o777)
            """)

def test_complete_two_context(client):
    conanfile = textwrap.dedent("""
        from conan import ConanFile

        class Pkg(ConanFile):
            name = "pkg"
            version = "0.1"
            apply_env = False

            def requirements(self):
                self.requires("openssl/1.0")
            def build_requirements(self):
                self.build_requires("openssl/1.0")

            def build(self):
                self.run("myopenssl.bat")
       """)

    client.save({"conanfile.py": conanfile})
    client.run("create . -pr:b=default -pr:h=default --build=missing")
    assert "MYOPENSSL=Windows!!" in client.out
```

And it works without issues in Conan 1.60. As you can see, the ``openssl`` recipe doesn't need to add anything to the path explicitly. Could you please try to provide something that we could reproduce? Thanks!

**Comment by db4** [comment]
I forgot to mention that I used `conan create --build-require ...` (I create a package for a tool that uses another tool). Add this flag to your example and it will fail.

**Comment by memsharded** [comment]
Ok, that helps.

There is indeed a limitation in the graph model in Conan 1.X, that it is not able to correctly compute the transitive dependencies of ``tool_requires`` that in turn have ``tool_requires`` and ``requires`` to the same dependency. (By the way, it would be interesting to know what exact dependencies you have to have that setup, it is not very common).

The result is that the transitive dependencies collapse into 1, in the "host" context of the dependency (which is the "build" context of the consumer, because of ``--build-require``). The workaround would be to explicitly use that context in the ``run``, like ``self.run("myopenssl.bat", env=["conanbuild", "conanrun"])``:

```python
def test_complete_two_context(client):
    conanfile = textwrap.dedent("""
        from conan import ConanFile

        class Pkg(ConanFile):
            name = "pkg"
            version = "0.1"
            apply_env = False
            settings = "os"

            def requirements(self):
                self.requires("openssl/1.0")

            def build_requirements(self):
                self.build_requires("openssl/1.0")

            def build(self):
                self.run("myopenssl.bat", env=["conanbuild", "conanrun"])
       """)

    client.save({"conanfile.py": conanfile})
    client.run("create . -pr:b=default -pr:h=default --build=missing --build-require")

    assert "MYOPENSSL=Windows!!" in client.out
```

**Comment by db4** [comment]
> By the way, it would be interesting to know what exact dependencies you have to have that setup, it is not very common

My case is indeed not quite common (another language compiler installer that uses a bunch of other tools to build and run), but imagine a simplified one: a tool that uses say Cygwin to build itself and also requires some Cygwin utilities (cp, mv, rm, etc.) during runtime.

> self.run("myopenssl.bat", env=["conanbuild", "conanrun"])

Thanks, that seems to work.

BTW, to target `x86` while creating a build-require package on `x86_64` host I have to use the following hack:
```python
    def generate(self):
        if self.settings.os == "Windows":
            tc = VCVars(self)
            if getattr(self, 'settings_target', None):
                self.settings_save = self.settings
                self.settings = self.settings_target
                tc.generate()
                self.settings = self.settings_save
                delattr(self, 'settings_save')
            else:
                tc.generate()
```
Is it safe in your opinion?

**Comment by memsharded** [comment]
> BTW, to target x86 while creating a build-require package on x86_64 host I have to use the following hack:

I am not sure about this approach. In theory that ``VCVars`` is being activated to build the current application, and the current application (tool-require) will run on x86_64. The ``target`` would be the code compiled with the current application, but that doesn't depend on the current ``VCVars``, but it should be something internal to its code, maybe controlled by some pre-processor directive, compile option or the like. Maybe what you are doing is that by forcing the current application to be built for ``x86``, it also forces that it will generate/compile code to ``x86``?

**Comment by db4** [comment]
Ok, here are some details. I am building an [OCaml](http://www.ocaml.org) compiler environment that produces Windows x86 native code binaries. To compile the compiler itself MSVC x86 toolchain is needed. The host OS is always Windows x86_64. So I do 

```
conan create . -pr:b default -s compiler.version=191 -s arch=x86 --build-require
```
The recipe is created according to [Cross building with Conan](https://docs.conan.io/1/systems_cross_building/cross_building.html?highlight=cross#conan-v1-24-and-newer) guidelines - the settings are encapsulated into a dedicated option and removed from package_id:
```python
    def package_id(self):
        del self.info.settings.compiler
        del self.info.settings.build_type
        ...
```
so the package can be used on x86_64 host as a build_require.

**Comment by memsharded** [comment]
This is what I was wondering. As the "build" and "host" for this package seems to be ``x86``, maybe the command line should be:
```
conan create . -pr:b default -s compiler.version=191 -s:h arch=x86 -s:b arch=x86 --build-require
```
Basically, the forcing of the "build" to ``x86`` doesn't need to be in the recipe, it can be in the command line.

**Comment by db4** [comment]
I need some tools that are x86_64-only. For instance, `msys2/cci.latest` just cannot be built for x86. And actually, an extra x86 binary is not needed, again the host is always x86_64.

**Comment by memsharded** [comment]
The
```
conan create . -pr:b default -s compiler.version=191 -s:h arch=x86 -s:b &:arch=x86 --build-require
```

Will only apply the build setting to the currently created package (``ocaml``), but not for the dependencies

**Comment by db4** [comment]
No, `-pr:b default --build missing -s:b "&:arch=x86" -s:h arch=x86 --build-require` leads to
```
Configuration (profile_host):
[settings]
arch=x86
arch_build=x86_64
build_type=Release
compiler=Visual Studio
compiler.runtime=MD
compiler.version=15
os=Windows
os_build=Windows
[options]
[build_requires]
[env]

Configuration (profile_build):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=Visual Studio
compiler.runtime=MD
compiler.version=15
os=Windows
os_build=Windows
&:arch=x86
[options]
[build_requires]
[env]
```
but VCVars() still sets up x86_64 environment (Conan 1.x)

**Comment by AbrilRBS** [comment]
Closing as we didn't hear back in a while, and this is Conan 1 related
